### DISCUSSION TITLE
Timer not working/always resetting?

### DISCUSSION THREAD

**Discussion by CameraRick** [arendst/Tasmota#22420] [Q&A] [answered]
Hi there,

I wanted to implement a timer, pretty much exactly [this one](https://github.com/arendst/Tasmota/discussions/13314). I have used it in the past, and it worked great; now I want to re-implement it, and it doesn't work anymore.
I tried this around two weeks ago, I installed the newest firmware _14.2.0(release-tasmota)_, I even reset the whole plug. Nothing helped, ran out of time, and today I'm looking at it again.

Rule setup is the following:
```
Mem1 20
Mem2 40
Mem3 300

Rule1 on Energy#Power<%Mem1% do ruletimer1 %Mem3% endon on Energy#Power>%Mem2% do ruletimer1 0 endon
Rule1 1
Rule1 5

Rule2 on rules#timer=1 do power off endon
Rule2 1
```

So basically, "if wattage is under 20 trigger the shutdown-timer of 5min, if wattage rises over 40 again stop the timer"

This worked exactly once, the first time I put in the rules, and then never again.
For debugging, I disabled the `Once` feature by doing `Rule1 4`, and then something confusing pops up:

```
14:03:29.999 MQT: stat/tasmota_1C939B/RESULT = {"T1":300,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
14:03:30.989 RUL: ENERGY#POWER<%MEM1% performs 'ruletimer1 300'
14:03:31.000 MQT: stat/tasmota_1C939B/RESULT = {"T1":299,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
14:03:31.007 RUL: ENERGY#POWER<%MEM1% performs 'ruletimer1 300'
14:03:32.001 MQT: stat/tasmota_1C939B/RESULT = {"T1":299,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
14:03:32.992 RUL: ENERGY#POWER<%MEM1% performs 'ruletimer1 300'
14:03:33.001 MQT: stat/tasmota_1C939B/RESULT = {"T1":299,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
14:03:33.949 RUL: ENERGY#POWER<%MEM1% performs 'ruletimer1 300'
14:03:33.959 MQT: stat/tasmota_1C939B/RESULT = {"T1":300,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
14:03:34.987 RUL: ENERGY#POWER<%MEM1% performs 'ruletimer1 300'
14:03:34.997 MQT: stat/tasmota_1C939B/RESULT = {"T1":300,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
14:03:35.987 RUL: ENERGY#POWER<%MEM1% performs 'ruletimer1 300'
14:03:35.996 MQT: stat/tasmota_1C939B/RESULT = {"T1":299,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
```

I'm no expert, not even an apparatentice by any means, but to me it looks like the timer always resets itself again. So it starts at 300 seconds, counts down to 299, and gets back to 300 again. And this is an endless loop.
So I suspect there must be some kind of error in the setup; I don't understand much of coding, and as I only copy+pasted this proven command I can't for the life of me figure out what's wrong. Maybe someone of you may know? :O

**Comment by barbudor** [comment]
This is where the Once mode helps by not triggering `Energy#Power<%Mem1%` continuously while the energy is below the lower threshold

**Reply by CameraRick** [reply]
Ah, so I misinterpret the output :(

I had deactivated the rules again after this post (I wanted to make a firmware upgrade and didn't want to risk it suddenly working and switching off the device). Now I re-activated them after I read your post, and suddenly it switched off as intended.
To make sure it actually works I switched it on again (10W load) and... nothing. Won't turn it off :(

**Reply by barbudor** [reply, marked as answer]
Because you designed your rule to start the timer on the energy going from above 20W to below 20W
If it stays at 10W from the power on, the event triggering the timer will not happen

In your case, a solution would be to start the timer on POWER ON so if the power doesn't go above 40W, it will POWER OFF as the timer reach 0

**Reply by CameraRick** [reply]
Hm, that is interesting.
As I understand the rule, it just says "if Power IS less than 20W", not "if Power BECOMES less than 20"; how can I know when such a condition has to be over the given limit first? I will test this now by quickly getting it to over 40W :)

**Reply by barbudor** [reply]
Indeed
Without Once it is "if Power IS less than 20W"
With once it is "if power goes from above 20W to below 20W"

But as you noticed, without Once, it would continuously restart the timer

So you have to move from a state based logic to an event based logic

**Reply by CameraRick** [reply]
Ah, I understand! That makes sense.
As I also use this for my 3D printer, the usual case would always be that it passes the threshold by drawing a lot more than 20W, for my testing I just only switched the unit on. Now for tests, I just set the heatbed to 50°C for a few secs where it draws over 1kW, then just stop it again. Works as it should in all tests I pulled.

Thanks for your swift help and explanations @barbudor it's so much appreciated :)
